Separate shared/service workers from the main page

ulan opened the pull request:
This applies @domenic's suggestion from
https://github.com/WICG/performance-measure-memory/issues/4#issuecomment-594057671

ulan commented:
@domenic, @annevk: please take a look.

domenic commented on README.md:
```diff
@@ -149,6 +149,11 @@ Adding a `userAgentSpecific` prefix to the `bytes` and `attribution` fields woul
 
+### Shared Workers and Service Workers
+Shared/service workers have their own JS agent clusters and do not belong to any browsing context group.
+This means that the result of an API call on the main page does not account shared/service workers.
```
"account shared/service workers" is a bit weird; "account for" or "count" maybe? Same below.

ulan replied:
I removed this sentence.

domenic commented on README.md:
```diff
@@ -87,7 +87,7 @@ Thus the two proposals are orthogonal.
 ## API Proposal
 The API consists of a single asynchronous method `performance.measureMemory` that estimates memory usage of the web page and provides breakdown of the result by type and owner.
-More formally, the API estimates memory usage of all [JS agent clusters](https://html.spec.whatwg.org/multipage/webappapis.html#integration-with-the-javascript-agent-cluster-formalism) of the current [browsing context group](https://html.spec.whatwg.org/multipage/browsers.html#browsing-context-group) including shared resources such as ServiceWorkers, SharedWorkers, SharedArrayBuffers.
+More formally, the API estimates memory usage of all [JS agent clusters](https://html.spec.whatwg.org/multipage/webappapis.html#integration-with-the-javascript-agent-cluster-formalism) of the current [browsing context group](https://html.spec.whatwg.org/multipage/browsers.html#browsing-context-group).
```
I know this is just the intro paragraph, but this sentence doesn't really make sense for the worker context. Maybe add something like "(In the shared/service worker case, it estimates memory usage of the worker's JS agent cluster"?)

ulan replied:
Moved the scope definitions into a separate section below.

ulan commented:
Thanks @domenic!

I uploaded a new patch that addresses the comments. I think it is better to disallow the API in dedicated workers because they can be spawned by iframes and other workers. Otherwise we would have to allow them conditionally (i.e. whitelist workers spawned by the main window) and explain what they return. The API call from the main window will measure all dedicated workers, so there is no need to call the API from dedicated workers.

domenic commented:
Hmm, I'm not sure I follow the reasoning. Why not allow it in dedicated workers, and it gives the surrounding BCG/agent cluster? You could make that work in dedicated workers spawned by shared/service workers too; they'd just report the result for the shared/service worker agent cluster.

ulan commented:
iframes are not allowed to call the API due to COOP. If we allow the API in workers, an iframe could spawn a worker and get the results. So we have to disallow workers spawned by iframes.

With this conditional availability the usage of API gets complicated. Besides that, if the user calls the API both from the main window and from a worker, then they have to do something to avoid double counting. This is not a problem for the main window and shared worker because the results are disjoint.

domenic commented:
Got it, that makes sense!

annevk commented:
An iframe that's cross-origin isolated will only spawn a worker that's cross-origin isolated. If the iframe is not cross-origin isolated, it cannot create a worker that is. So I don't think it makes sense, but I might be missing something.

ulan commented:
@annevk that is the conditional availability: workers spawned by some iframes and the main window will get the API and others will not get. We can specify that using `crossOriginIsolated` flag, but my main argument is that the API will be easier to use (and harder to misuse) if we allow it only for the main window and shared/service workers. Then there is no risk of double counting by calling the API in iframes and workers.
Since the main window call will measure all nested context, there is no need to allow API in nested contexts.

annevk commented:
I still don't understand, it's not possible to have an iframe that has cross-origin isolated in a different state from its parent, siblings, or children.

ulan commented:
Maybe I am confused, but I think that cross-origin iframes embedded in a cross-origin isolated page have to have their `window.crossOriginIsolated = false`. Otherwise, they can use SharedArrayBuffers and the memory API to leak information from the main page.

In general, since COOP is ignored for iframes, doesn't it mean that `crossOriginIsolate = false` for all iframes?

annevk commented:
No it doesn't. If they don't opt-in to COEP they don't load. COOP+COEP is a per browsing context group affair, it's not per agent cluster.

(And yes, the parent page has to trust them, just as they have to trust the parent page. The whole ceremony is about mutual trust.)

ulan commented:
Thanks for explaining! This is new to me. I always thought that the trust is in only in one direction: iframes trust the main page.

Even if `crossOriginIsolated` is the same for all browsing contexts, I would like to restrict the API only to the top level contexts. Now for two reasons: 1) usability (avoiding double counting) 2) security  (so that cross-origin iframe cannot measure memory of the main page).

Does this sound good?

annevk commented:
I don't have all the context, but generally I would expect it to be a first-party/top-level origin restriction in that case. E.g., if you happen to run some of your code in a same origin frame, it should be able to do this.

ulan commented:
To give more context: the intended use of the API is to have a global memory monitor for a page that performs statistical sampling of memory usage and sends the samples to the server for A/B testing and regression detection. The API is not suitable for local measurements such as measuring the memory before and after some action.

Calling the API from within an iframe is likely a misuse: if the main page is already sampling the memory usage, then an iframe sampling the same data would introduce a bias (double counting).

I see that it is desirable to keep the code consistent between the main frame and same-origin iframes for synchronous scripting. But if an iframe is synchronously calling memory measuring code of the main frame, then something is already off with the usage of the API.

If the consistency is not a hard spec requirement, my preference would be to forbid the calls in iframes.

annevk commented:
How can you meaningfully forbid them calling it though? There's all kinds of ways they would be able to get to it if there's script access.

ulan commented:
By "forbidding" here I mean that the API would throw an error if it is invoked in an iframe context.
(In the case of a cross-origin iframe, the API will definitely throw for security reasons.)

ulan commented:
I think I see what you mean. Throwing meaningful only for the direct calls of the API. If the API is wrapped in another function from the main page, then iframe would be able to call it.

annevk commented:
And the frame has the ability to create such functions too.

ulan commented:
Thanks a lot for your input here! It was helpful.

I agree that same-origin iframes should have the same access to the API as the main window. Restricting the access is not enforceable and leads to inconsistencies. I'll update the PR.

We still have a choice for dedicated workers, where the restriction is enforceable.
Allowing the API would make it consistently available in all same-origin contexts.
But as mentioned before, there is no good reason for a user to call the API from a dedicated worker. The conditional availability based on the origin may be confusing for the users.
There is also implementation cost: it would have to handle concurrent memory measurement requests from multiple JS agents within a JS agent cluster. Any preferences here?

annevk commented:
I think it's okay to restrict it to the "main thread" agent for that reason. (Though it's also unfortunate we keep adding (costly?) main thread APIs. cc @surma)

ulan commented:
Keeping track of the pending requests becomes either single-threaded or multi-threaded depending on the decision about the workers.

The actual implementation is either cheap (reading memory counters) or can be offloaded to a background thread (e.g. performed with a concurrent GC).

ulan commented:
@annevk, @domenic does https://github.com/WICG/performance-measure-memory/pull/9/commits/49bf180f0f9978c5d963346c536b9bd844a4ce64 look good to you?

domenic commented on README.md:
````diff
@@ -171,6 +182,8 @@ try {
 }
 ```
 
+Invoking the API in a cross-origin iframe also throws a `SecurityError` DOM exception even if `crossOriginIsolated === true`.
````
It isn't very clear what this means. Here is my best guess: "invoking the API in an iframe which is cross-origin from its parent". Or, do you check the entire ancestor tree?

ulan replied:
Thanks. I meant cross-origin from the main window. Clarified that in the text.

Checking for origin relative to the parent would allow the following scenario:
1) foo.com embeds bar.com/iframe1 which embeds bar.com/iframe2. 
2) bar.com/iframe2 would be allowed to measure memory which may leak info from foo.com.
